### ISSUE TITLE
Feature Request: Spotlight Tool

### ISSUE THREAD

**Issue by rexelbartolome** [pubpub-zz/ppInk#135]
Think as if when you select this **Spotlight Tool** or you're in **Spotlight Mode**, you start with a `50% opaque` or `120 alpha` black color all over the screen but when you use one of the shape tools, instead of drawing a shape on top of the dark overlay, it then erases a part of the that dark overlay and you're left with a spotlight effect highlighting a specific area in the screen. 

`Select Spotlight Mode > Select Rectangle Tool > Draw`

![image](https://user-images.githubusercontent.com/88552647/128989480-c2b329e6-99e4-40ec-bcf0-b020acfd7486.png)

It can also have a Fading effect similar to the existing Fading feature where the drawn shape disappears after a few seconds.

Another feature is just a copy of a similar tool of another program that I already use but is based on .ahk and not .cs **Cursor Highlight 2.2 AHK**. It just adds a spotlight around the cursor. I guess you can add this when you select the Cursor Tool after the Spotlight Tool.

`Select Spotlight Mode > Select Cursor Tool`

![image](https://user-images.githubusercontent.com/88552647/128989517-487d0bc0-0bc3-4e0c-bea1-b49ae5120dfe.png)

I honestly just want to have a Rectangle Shape Spotlight, I wouldn't really use Ellipse nor Freehand as much, at least for my purposes. 

I also do understand that making this feature start of as a Mode, then making all the drawing tools react differently than normal might limit others since other people might want to draw on the screen while the spotlight is active so maybe a setup like this might be more versatile

`Rectangle Tool > Spotlight Option`

![image](https://user-images.githubusercontent.com/88552647/128991524-e4c81c5c-a8f4-4335-990f-c04e08a7ab45.png)

Anyway, I'll leave it up to you _if_ or _how_ you'll go about adding this feature 😊

Icon was from [Flaticon](https://www.flaticon.com/free-icon/flashlight_4290763?term=flashlight&page=1&position=15&page=1&position=15&related_id=4290763&origin=search)

**Comment by pubpub-zz** [comment]
Filling outside of an area, I see a solution for rectangle only ( I will plot 4 Rectangles outside leaving the inner empty. I will put that in the stack but sounds feasable. It will be within the rectangles.

About the Spot following the cursor, I'm not sure yet I can have a round spot light :  it may be also a rectangle. I will probably add the tool within the zoom button to. I will have also to assert to see how performances can match this ( with wide surface from multiple monitors)
I will for the moment leave that with "feasability to be analysed" to track it the task list

**Comment by rexelbartolome** [comment]
Glad to hear it sounds feasable! Looking forward to it 😊

**Comment by pubpub-zz** [comment]
@rexelbartolome ,

I've just finished a first version with outside filling : I've been able to apply it all tools.
[ppInk64_Outside17b.zip](https://github.com/pubpub-zz/ppInk/files/6999505/ppInk64_Outside17b.zip)

Spotlight on Cursor should come soon 😉

**Comment by rexelbartolome** [comment]
OMG That's so cool!!! Thanks so much 😊

Can't believe you managed to make it work for all of them! 🥳🎉 I'll be definitely using the rectangular one, and I'll be recommending this program to others as well! (ofc once it's merged with the main branch or something?)

I guess the only thing that I can complain about is the fact that you need to press the same tool 4 times

**Comment by rexelbartolome** [comment]
Also I didn't mean to close it, and I'm not sure how to reverse it? 🤐

**Comment by pubpub-zz** [comment]
@rexelbartolome 
> I guess the only thing that I can complain about is the fact that you need to press the same tool 4 times to get to the Outside mode, tho I think it saves the mode if you turn ppInk off through Ctrl+Alt+G, and not using Escape? So I'll be figuring that one out :)
I Moved it to just after Pen Color, However,normally you will use this filling only once every drawing session so it should not be a problem.
I've also finished the spot. You can activate it in 2 ways : reaching the tool at the end of zoom,  and depressing Alt (can be switch off in options)
Also, color/transparency and radius can be changed in options.
[ppInk64_Spot17c.zip](https://github.com/pubpub-zz/ppInk/files/7003256/ppInk64_Spot17c.zip)

@NOVAinc, you may be interested to have a look at those two new features

**Comment by rexelbartolome** [comment]
Awesome! :D Glad it's finally here 😊

Yea, going to the Outside mode is a bit easier now. Tho you just need to remember how many times you pressed it. I also just figured out ppInk remembers the last tool and mode selected if it was deactivated by going to the Pointer mode but it goes back to the default Freehand if it was deactivated by pressing Escape. 

I guess since I'm using this as a real-time teaching tool, and less so of a screenshot annotator, I just expected adding a editable shortcut option for the specific mode I want for quicker access, but that would also mean, to be consistent across all tools, you'd need to add ALL of the tool's modes, which would mean that you might need to redo the entire Hotkey tab GUI 😅 So I think this implementation will work well for me for now :D I'm just glad that this is all possible.

Also, the default binding of the Global Hotkey for activating/deactivating ppInk (Ctrl+Alt+G) also had a weird interaction with the Spotlight cursor. When deactivating or entering Mouse Pointer mode, the Spotlight gets activated and doesn't go away, which I think is unintentional, since going to the Mouse Pointer mode manually by clicking on it on the Docker wouldn't have the same result. Same goes for engaging pointer and auto-fold with Alt-Tab.

Also let me know what can I do to help regarding UI and stuff, since I'm an artist and graphic designer as well, don't know how to code in C# unfortunately 😅 so I can only provide the assets and mockups but not personally implement them...

**Comment by pubpub-zz** [comment]
> Also, the default binding of the Global Hotkey for activating/deactivating ppInk (Ctrl+Alt+G) also had a weird interaction with the Spotlight cursor. When deactivating or entering Mouse Pointer mode, the Spotlight gets activated and doesn't go away, which I think is unintentional, since going to the Mouse Pointer mode manually by clicking on it on the Docker wouldn't have the same result. Same goes for engaging pointer and auto-fold with Alt-Tab.

The issue was about the Spot on Alt that is not unset as when entering Pointer Mode. This should be fixed:
[ppInk64_Spot18a.zip](https://github.com/pubpub-zz/ppInk/files/7005886/ppInk64_Spot18a.zip)
I've also added the REST API control for it.
However if you want to keep the Spot mode active in pointer mode you can still activate it under the zoom button which I found quite usefull.

> Also let me know what can I do to help regarding UI and stuff, since I'm an artist and graphic designer as well, don't know how to code in C# unfortunately 😅 so I can only provide the assets and mockups but not personally implement them...

Thanks for your proposal. I will keep it in mind. Esthetic may not be my biggest strength but it is not always easy to have an OMI that matches all needs, and also I want to be sure that icons' license will be compatible with ppInk's.

**Comment by rexelbartolome** [comment]
The Spotlight cursor is working as intended! Thank you 😊

Tho after transferring my settings to this build, I noticed a bunch of things

Changing the pen color inside the Settings>Pen will change the width and alpha values with it for some reason

Disabling the clip art checkbox will make the toolbar cut off at the midpoint of the Close icon
![image](https://user-images.githubusercontent.com/88552647/130048027-11c03fe3-44da-45fb-b88d-c55c918aa0cd.png)

But also dragging the toolbar doesn’t cause weird issues anymore :D

Also what do you mean by OMI's?

**Comment by pubpub-zz** [comment]
> Changing the pen color inside the Settings>Pen will change the width and alpha values with it for some reason

Got it ; under analysis but will be fixed soon
> 
> Disabling the clip art checkbox will make the toolbar cut off at the midpoint of the Close icon
> ![image](https://user-images.githubusercontent.com/88552647/130048027-11c03fe3-44da-45fb-b88d-c55c918aa0cd.png)

found : will be fixed soon

> But also dragging the toolbar doesn’t cause weird issues anymore :D> 

> Also what do you mean by OMI's?
Operator Machine Interface ; It's the word I learnt 20 years ago... 😏 I should use  with UI to be more modern...👴😀

**Comment by rexelbartolome** [comment]
Lol, never heard of that OMI term before 😂 I only know of GUI, UI, or HUD.

Anyway, I have some thoughts on how to improve some of the icons and UI, mainly the icons that have aliasing and a bit far in style from the rest. And I'm not sure how to contact you directly, since I have some questions regarding what the icon's license should be for it to be compliant with ppInk's 🤔 can we talk in Discord? My tag is Rexel#6689

**Comment by pubpub-zz** [comment]
> 
> 
> > Changing the pen color inside the Settings>Pen will change the width and alpha values with it for some reason
> 
> Got it ; under analysis but will be fixed soon
> 
> > Disabling the clip art checkbox will make the toolbar cut off at the midpoint of the Close icon
> > ![image](https://user-images.githubusercontent.com/88552647/130048027-11c03fe3-44da-45fb-b88d-c55c918aa0cd.png)
> 
> found : will be fixed soon
> 
> > But also dragging the toolbar doesn’t cause weird issues anymore :D>
> 
> > Also what do you mean by OMI's?
> 
> Operator Machine Interface ; It's the word I learnt 20 years ago... 😏 I should use with UI to be more modern...👴😀

@rexelbartolome,
bugs are fixed in 
[ppInk64_Arrow21a.zip](https://github.com/pubpub-zz/ppInk/files/7025950/ppInk64_Arrow21a.zip)

**Comment by rexelbartolome** [comment]
> @rexelbartolome,
> bugs are fixed in
> [ppInk64_Arrow21a.zip](https://github.com/pubpub-zz/ppInk/files/7025950/ppInk64_Arrow21a.zip)

Cool! Tested it out and it's working great :)

Tho where do I find more information regarding the compatible license for ppInk? I saw that both gInk's and ppInk's used the icon in this one https://www.flaticon.com/ru/free-icon/rubber_65602 for the eraser which has an attribution license. So if attribution licenses are compatible with this project's license (MIT) then that would make the whole thing easier :D Let me know if I should make a separate issue regarding this or maybe ask geovens instead.
